### PR TITLE
streamingccl: support protected timestamp in stream replication producer job

### PR DESCRIPTION
**Pull request by gh-casper**
Add protected timestamp into producer job: create a pts record when a producer job is created, update it when a frontier is updated, release it when a job failed.

Release note: none

### CODE REVIEW

**Review by gh-casper — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72404)* status: :shipit: complete! 0 of 0 LGTMs obtained

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-Mo9n8ax9h0XETZVVldw:-Mo9n8ax9h0XETZVVldx:boig35t) (no related file):*
Can you just review the second commit? The first commit is the ongoing "create_replication_stream" change.

---

File: pkg/ccl/streamingccl/streamproducer/producer_job.go [hunk 1]
```diff
@@ -105,13 +131,73 @@ func doStartReplicationStream(
...
+	ptp := execConfig.ProtectedTimestampProvider
+	statementTime := hlc.Timestamp{
+		WallTime: evalCtx.GetStmtTimestamp().UnixNano(),
+	}
+	pts := jobsprotectedts.MakeRecord(ptsID, int64(jr.JobID), statementTime,
+		[]roachpb.Span{*makeTenantSpan(tenantID)}, jobsprotectedts.Jobs)
+
+	if err := ptp.Protect(evalCtx.Ctx(), txn, pts); err != nil {
+		return streaming.InvalidStreamID, err
+	}
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamproducer/producer_job.go:151]: I'm pretty sure we want to create PTS record before we start the job.  E.g. see changefeed_stmt.
Basically, we don't ever want to run streaming w/out knowing we have protected data.

File: pkg/ccl/streamingccl/streamproducer/producer_job.go [hunk 2]
```diff
@@ -105,13 +131,73 @@ func doStartReplicationStream(
...
+
+	d := jr.Details().(jobspb.StreamReplicationDetails)
+	ptsRecord, err := ptsProvider.GetRecord(ctx, txn, *d.ProtectedTimestampRecord)
+	if err != nil {
+		return err
+	}
+
+	if shouldUpdatePTS := ptsRecord.Timestamp.
+		Add(streamingccl.MinProtectedTimestampUpdateAdvance.Nanoseconds(), 0).
+		Less(ts); shouldUpdatePTS {
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamproducer/producer_job.go:187]: Maybe this MinProtectedTS advance is .. a bit premature optimization.
Perhaps we can simply update pts record to ts provided ts > ptsRecord.Timestamp?
We can add min advance later.

File: pkg/ccl/streamingccl/streamproducer/producer_job.go [hunk 3]
```diff
@@ -105,13 +131,73 @@ func doStartReplicationStream(
...
+		if err = ptsProvider.UpdateTimestamp(ctx, txn, *d.ProtectedTimestampRecord, ts); err != nil {
+			return err
+		}
+	}
+
+	return registry.UpdateJobWithTxn(ctx, jobspb.JobID(streamID), txn, false,
+		func(txn *kv.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamproducer/producer_job.go:194]: I wonder if the entirety of the above code must happen here, inside UpdateJobWithTxn?

For one, if txn gets retried, you get up to date job record.  So, your checks whether you need to update will
be more accurate.  Secondly, registry takes care of making sure the job is running, so I think you'd be able to drop some of the checks you have above.  Finally, in the happy case, 1 less read against jobs table.

File: pkg/ccl/streamingccl/streamproducer/producer_job_test.go [hunk 4]
```diff
@@ -148,45 +144,116 @@ func TestStreamReplicationProducerJob(t *testing.T) {
...
+			ptsID := uuid.MakeV4()
+			jr := makeProducerJobRecord(registry, 10, timeout, username, ptsID)
+			defer jobs.ResetConstructors()()
+			_, _, _, waitUntilReverting := registerConstructor(expirationTime(jr).Add(1 * time.Millisecond))
+
+			require.NoError(t, runJobWithProtectedTimestamp(ptsID, ts, jr))
+
+			waitUntilReverting()
+
+			sql.SucceedsSoonDuration = 3 * time.Second
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamproducer/producer_job_test.go:184]: I will keep asking the same question: why do we need to override this?

**Review by miretskiy — CHANGES_REQUESTED** [review]

**Comment by miretskiy** [comment]
Okay.

**Review by gh-casper — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72404)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-Mo9n8ax9h0XETZVVldw:-MoAms_LES2C3CVBaYXK:bvfkwpx) (no related file):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

Can you just review the second commit? The first commit is the ongoing "create_replication_stream" change.
</blockquote></details>

I rebased it on the current master after that PR got merged.

---
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 151 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-MnbDHwW81CAj-gR-OJj-r5-151:-MoAPYJRC8dZS5xz9olX:b-k36k8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/788640dc08da5b4db318d3ece6c1d321c21ed8eb/pkg/ccl/streamingccl/streamproducer/producer_job.go#L151)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm pretty sure we want to create PTS record before we start the job.  E.g. see changefeed_stmt.
Basically, we don't ever want to run streaming w/out knowing we have protected data.

</blockquote></details>

They will be running in the same transaction (txn object passed in from doStartReplicationStream function). No?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 187 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-MnbDHwW81CAj-gR-OJj-r5-187:-MoASWaIB3a_Y8UskQSd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/788640dc08da5b4db318d3ece6c1d321c21ed8eb/pkg/ccl/streamingccl/streamproducer/producer_job.go#L187)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Maybe this MinProtectedTS advance is .. a bit premature optimization.
Perhaps we can simply update pts record to ts provided ts > ptsRecord.Timestamp?
We can add min advance later.
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 194 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-MnbDHwW81CAj-gR-OJj-r5-194:-MoAS6cSBUgNn0oayH_F:bw8ciz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/788640dc08da5b4db318d3ece6c1d321c21ed8eb/pkg/ccl/streamingccl/streamproducer/producer_job.go#L194)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I wonder if the entirety of the above code must happen here, inside UpdateJobWithTxn?

For one, if txn gets retried, you get up to date job record.  So, your checks whether you need to update will
be more accurate.  Secondly, registry takes care of making sure the job is running, so I think you'd be able to drop some of the checks you have above.  Finally, in the happy case, 1 less read against jobs table.
</blockquote></details>

Same comment above. This `UpdateJobWithTxn` shares the same `txn` object with `provider.UpdateTimestamp()`. So they are running in the same txn?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 184 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-MnbDHwXH6my1qpEjQSK-r5-184:-MoASUJrEEz9CYELA33C:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/788640dc08da5b4db318d3ece6c1d321c21ed8eb/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L184)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I will keep asking the same question: why do we need to override this?
</blockquote></details>

Removed.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 215 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-MnbDHwXH6my1qpEjQSK-r5-215:-MoAR5AL3Xibc7PBwr64:b-e7rdn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/788640dc08da5b4db318d3ece6c1d321c21ed8eb/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L215)):*
> waitUntilReverting
I should have renamed `waitUntilReverting` to be `waitJobFinishReverting` because the channel is actually signaled after `OnFailOrCancel` runs. In this case, we are sure that the job reaches "reverting" stage after `waitJobFinishReverting()`.

**Review by miretskiy — APPROVED** [review]
> Reviewed 1 of 22 files at r1, 2 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72404)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @gh-casper, and @miretskiy)

___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 151 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-MnbDHwW81CAj-gR-OJj-r5-151:-MoB-h7BCReIpDiIDwM-:b-xd3vm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/788640dc08da5b4db318d3ece6c1d321c21ed8eb/pkg/ccl/streamingccl/streamproducer/producer_job.go#L151)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

They will be running in the same transaction (txn object passed in from doStartReplicationStream function). No?
</blockquote></details>

You're right.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 194 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-MnbDHwW81CAj-gR-OJj-r5-194:-MoB09Wv5aG4L1WwuJwS:brol7hq) ([raw file](https://github.com/cockroachdb/cockroach/blob/788640dc08da5b4db318d3ece6c1d321c21ed8eb/pkg/ccl/streamingccl/streamproducer/producer_job.go#L194)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

Same comment above. This `UpdateJobWithTxn` shares the same `txn` object with `provider.UpdateTimestamp()`. So they are running in the same txn?
</blockquote></details>

Yes, you're correct that it's the same txn... But UpdateJob will execute select to load job record -- so, at the very least, let's not
do this twice?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 215 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72404#-MnbDHwXH6my1qpEjQSK-r5-215:-MoB0Prh97UG3kSy0zh1:b-wh3mqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/788640dc08da5b4db318d3ece6c1d321c21ed8eb/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L215)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

> waitUntilReverting
I should have renamed `waitUntilReverting` to be `waitJobFinishReverting` because the channel is actually signaled after `OnFailOrCancel` runs. In this case, we are sure that the job reaches "reverting" stage after `waitJobFinishReverting()`. 
</blockquote></details>

I see.  Makes sense.  Still , double check ? will you?

**Pull request merged**